/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/GUIForms/JPanel.java to edit this template
 */
package GVJ.gui;

import static GVJ.utils.GffUtils.getSelectedFeatureLocations;

import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.geom.AffineTransform;
import java.io.Console;
import java.util.List;

import javax.swing.JOptionPane;

import org.biojava.nbio.genome.parsers.gff.FeatureList;
import org.biojava.nbio.genome.parsers.gff.Location;

import GVJ.models.DataManager;

import java.awt.Color;

/**
 *
 * @author mspriggs
 */
public class geneVisPanel extends javax.swing.JPanel {

    private DataManager dataManager;
    private String selectedFeature = "exon"; // Default feature type
    private String selectedGene;
    private FeatureList gff;
    private Boolean hasLabel = false;

    public void setHasLabel(Boolean hasLabel) {
        this.hasLabel = hasLabel;
    }

    /**
     * Creates new form geneVisPanel
     */
    public geneVisPanel() {
        initComponents();
        // Extract a list of Locations for each feature in gene
    }

    public void setDataManager(DataManager dataManager) {
        this.dataManager = dataManager;
    }

    public void setSelectedGene(String gene) {
        this.selectedGene = gene;
        repaint();
    }

    public void setSelectedFeature(String feature) {
        this.selectedFeature = feature;
        repaint();
    }

    /**
     * This method is called from within the constructor to initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is always
     * regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated
    // <editor-fold defaultstate="collapsed" desc="Generated
    // <editor-fold defaultstate="collapsed" desc="Generated
    // <editor-fold defaultstate="collapsed" desc="Generated
    // <editor-fold defaultstate="collapsed" desc="Generated
    // Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        setBorder(javax.swing.BorderFactory.createTitledBorder("Gene Visualization"));
        setMinimumSize(new java.awt.Dimension(250, 250));

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(this);
        this.setLayout(layout);
        layout.setHorizontalGroup(
                layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                        .addGap(0, 390, Short.MAX_VALUE));
        layout.setVerticalGroup(
                layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                        .addGap(0, 276, Short.MAX_VALUE));
    }// </editor-fold>//GEN-END:initComponents

    // Variables declaration - do not modify//GEN-BEGIN:variables
    // End of variables declaration//GEN-END:variables

    @Override
    public void paintComponent(Graphics g) {
        super.paintComponent(g);

        // Gene Representation
        int panelWidth = getWidth();

        int panelHeight = getHeight();

        int panelMidYCood = panelHeight / 2;

        int paddingPadding = panelWidth / 10;

        // Draw gene shape
        drawGeneShape(g, panelWidth, panelMidYCood, paddingPadding);

        // set highlight color
        Color highlightCol = setColor(selectedFeature);

        // Check if data is available
        if (dataManager == null) {
            // What should happen here?
            // - Draw nothing?
            // - Show a message?
            // - Just return early?
            return;
        }

        this.gff = dataManager.getGffData();

        try {

            List<Location> featureLocations = getSelectedFeatureLocations(gff,
                    selectedGene, selectedFeature);

            for (Location loc : featureLocations) {
                // get gene length
                List<Location> geneLocationList = getSelectedFeatureLocations(gff,
                        selectedGene, "gene");
                Location geneLoc = geneLocationList.get(0);

                // Map feature location to panel
                addFeatureRectangle(g, panelWidth, panelMidYCood, paddingPadding, highlightCol, loc, geneLoc);

            }

        } catch (NullPointerException e) {
            System.out.println("Error in paintComponent from geneVisPanel:");
            System.out.println(e.getMessage());

            return;
        }

    }

    private void drawGeneShape(Graphics g, int panelWidth, int midY, int padding) {

        // left of gene
        int geneLeftX = padding;

        // right of gene
        int geneRightX = panelWidth - padding;

        // top of gene
        g.drawLine(geneLeftX, midY - 15, geneRightX, midY - 15);
        // bottom of gene
        g.drawLine(geneLeftX, midY + 15, geneRightX, midY + 15);
        // left side of gene
        g.drawArc(geneLeftX - 15, midY - 15, 30, 30, 90, 180);
        // right side of gene
        g.drawArc(geneRightX - 15, midY - 15, 30, 30, 270, 180);
    }

    private void addFeatureRectangle(Graphics g, int panelWidth, int midY, int padding, Color highlightCol,
            Location featureLoc, Location geneLoc) {

        // left of gene drawing
        int geneLeftX = padding;

        // right of gene drawing
        int geneRightX = panelWidth - padding;

        // gene model Location for drawing
        Location geneDrawingLoc = new Location(geneLeftX, geneRightX);

        // relative gene drawing to gene model
        int featureRelativeStartBpCoor = featureLoc.start() - geneLoc.start();
        int featureRelativeEndBpCoor = featureLoc.end() - geneLoc.start();

        // get strand information
        char geneStrand = geneLoc.bioStrand();
        System.out.println("Gene Strand: " + geneStrand);

        // convert length in bp to pixels
        double scale = (double) (geneDrawingLoc.length()) / geneLoc.length();

        // find shape start and end in drawing coordinates
        int featureX1Coordinate = geneDrawingLoc.start() + (int) (featureRelativeStartBpCoor * scale);
        System.out.println("Feature X1 Coordinate: " + featureX1Coordinate);
        int featureX2Coordinate = geneDrawingLoc.start() + (int) (featureRelativeEndBpCoor * scale);
        System.out.println("Feature X2 Coordinate: " + featureX2Coordinate);
        int shapeWidth = featureX2Coordinate - featureX1Coordinate;

        // draw rectangle for feature
        int featureHeight = 30;
        g.setColor(highlightCol);

        // For positive strand, draw as is:
        g.fillRect(featureX1Coordinate, midY - 15, shapeWidth, featureHeight);

        // Add text label for feature
        if (!this.hasLabel) {

            String labelText = (featureLoc.bioStrand() == '+') ? "---> Forward" : "<--- Reverse";

            g.setColor(Color.BLACK);
            g.drawString(labelText + " Strand", featureX1Coordinate, midY - 20);
            this.hasLabel = true; // really poor way to track this but ok for noW #TODO refactor this whole
                                  // function. Ideally flip based on strand.
        }
    }

    // private void drawExon(Graphics g,
    private Color setColor(String featureType) {
        return switch (featureType) {
            case "mRNA" -> new Color(135, 206, 250, 150); // Light Blue
            case "intron" -> new Color(144, 238, 144, 150); // Light Green
            case "exon" -> new Color(255, 182, 193, 150); // Light Pink
            case "CDS" -> new Color(255, 255, 0, 150); // Yellow
            default -> new Color(211, 211, 211, 150); // Light Gray for unknown types
        };
    }

}
